ipywidgets/widgets/widget.py

```diff
@@ -317,6 +317,26 @@ def _call_widget_constructed(widget):
         if Widget._widget_construction_callback is not None and callable(Widget._widget_construction_callback):
             Widget._widget_construction_callback(widget)
 
+    @classmethod
+    def handle_comm_opened_control(cls, comm, msg):
+        cls.get_manager_state()
+        widgets = Widget.widgets.values()
+        # build a single dict with the full widget state
+        full_state = {}
+        drop_defaults = False
+        for widget in widgets:
+            full_state[widget.model_id] = {
+                'model_name': widget._model_name,
+                'model_module': widget._model_module,
+                'model_module_version': widget._model_module_version,
+                'state': widget.get_state(drop_defaults=drop_defaults),
+            }
+        full_state, buffer_paths, buffers = _remove_buffers(full_state)
+        # the message is also send as buffer, so it does not get handled by jupyter_server
+        msg = jsondumps([full_state, buffer_paths]).encode('utf8')
+        buffers.insert(0, msg)
+        comm.send(buffers=buffers)
```

Comment:
What is the reasoning behind this sending the entire state as a "buffer" instead of sending as regular JSON? Is it to sidestep the decoding that happens in the server?

Reply:
Yes. It has a big performance impact. Although I would agree we don't want to do that and we would want to improve the server instead.

Reply:
I'm not sure that is the right solution either, I think the fact that the jupyter protocol speaks JSON and the messages we send (the widget protocol) are JSON are a mere coincidence. We should not embed our widget protocol message inside the jupyter protocol as JSON, the JSON from the widget protocol should just be a string or binary buffer in the jupyter message. That way the server would not parse the JSON (it would just be a string). In this case I've chosen a buffer for performance reasons, but a single string would also suffice (i.e. json.dumps(..)). But I don't want to go too much off topic here.